iroh-net/src/hp/portmapper/upnp.rs

```diff
@@ -0,0 +1,119 @@
+use std::{
+    fmt::Display,
+    net::SocketAddrV4,
+    time::{Duration, Instant},
+};
+
+use anyhow::Error;
+use igd::aio as aigd;
+
+/// Seconds we ask the router to maintain the port mapping. 0 means infinite.
+const PORT_MAPPING_LEASE_DURATION_SECONDS: u32 = 0;
+
+/// Maximum duration a UPnP search can take before timing out.
+const SEARCH_TIMEOUT: Duration = Duration::from_millis(1000);
+
+const PORT_MAPPING_DESCRIPTION: &str = "iroh-portmap";
+
+#[derive(Debug, Clone)]
+pub struct Mapping {
+    /// The internet Gateway device (router) used to create this mapping.
+    gateway: aigd::Gateway,
+    /// The external address obtained by this mapping.
+    external_addr: SocketAddrV4,
+    /// Local address used to create this mapping.
+    local_addr: SocketAddrV4,
+    /// Instant in which the mapping was first created or last renewed.
+    created_at: Instant,
+}
+
+impl Mapping {
+    pub(crate) async fn new(local_addr: SocketAddrV4) -> Result<Self, Error> {
+        let gateway = aigd::search_gateway(igd::SearchOptions {
+            timeout: Some(SEARCH_TIMEOUT),
+            ..Default::default()
+        })
+        .await?;
+
+        let external_addr = gateway
+            .get_any_address(
+                igd::PortMappingProtocol::UDP,
+                local_addr,
+                PORT_MAPPING_LEASE_DURATION_SECONDS,
+                PORT_MAPPING_DESCRIPTION,
+            )
+            .await?;
+        Ok(Mapping {
+            gateway,
+            external_addr,
+            local_addr,
+            created_at: Instant::now(),
+        })
+    }
+
+    /// Releases the mapping.
+    pub(crate) async fn release(self) -> Result<(), Error> {
+        let Mapping {
+            gateway,
+            external_addr,
+            ..
+        } = self;
+        gateway
+            .remove_port(igd::PortMappingProtocol::UDP, external_addr.port())
+            .await?;
+        Ok(())
+    }
+
+    /// Renews the mapping and updates the external address (external ip could change).
+    pub(crate) async fn renew(&mut self) -> Result<(), Error> {
+        self.gateway
+            .add_port(
+                igd::PortMappingProtocol::UDP,
+                self.external_addr.port(),
+                self.local_addr,
+                PORT_MAPPING_LEASE_DURATION_SECONDS,
+                PORT_MAPPING_DESCRIPTION,
+            )
+            .await?;
+        let external_ip = self.gateway.get_external_ip().await?;
+        self.external_addr.set_ip(external_ip);
+
+        Ok(())
+    }
+
+    /// good_until will return the lease time that the mapping is valid for.
+    pub fn good_until(&self) -> Instant {
+        // assume an hour
```

Comment:
is this a documented value in the spec, or is this just what tailscale "figured out"

Reply:
their value is actually much higher, 2 hours, and supposedly comes from a recommendation from an RFC for another port mapping protocol. an hour seemed reasonable to me, but I can change it to the same 2 hours they use if that's the preference

Reply:
Happy to go with one hour for now, just add a comment where the number comes from please